### PR TITLE
[XLA:CPU] oneDNN MatMul constant weights pre-packing

### PR DESCRIPTION
**Pull request by rfsaliev**
This PR implements MatMul constant weights pre-packing(reordering) to optimized layout at compile time for performant MatMul execution at runtime. Non-constant weights kept unchanged to avoid buffers reordering at runtime.

As far as oneDNN MatMul weights optimization depends on runtime intra-op parallelization(concurrency) level, following changes made as well:
* Added static method to `OneDnnThreadPool` which allows to configure oneDNN parallelization level at compile time.
* PJRT CPU client and HLO runner modified to pass intra-op parallelization level to `HloModule` configuration.

To let weights pre-packing run faster at compile time, `compile_thread_pool` is used for `dnnl::reorder` calls, to manage this configuration properly following changes were made:
* `CpuCompiler::RunHloPasses()` signature is updated to receive `CompileOptions`.
* PJRT CPU client is modified to setup `CompileOptions::compile_thread_pool` for `CpuCompiler` calls.

### CODE REVIEW

File: xla/service/cpu/onednn_matmul.cc [hunk 1]
```diff
@@ -177,6 +278,93 @@ ABSL_ATTRIBUTE_NO_SANITIZE_MEMORY void __xla_cpu_runtime_OneDnnMatMul(
...
+          run_options->intra_op_thread_pool()->getPool(), false);
+      return stream(
+          dnnl::threadpool_interop::make_stream(cpu_engine, &thread_pool));
+    } else {
+      return stream(cpu_engine);
+    }
+  }();
+#else
+  auto onednn_stream = stream(cpu_engine);
+#endif  // ENABLE_ONEDNN_OPENMP
```
Comment (penpornk) [inline xla/service/cpu/onednn_matmul.cc:310]: This code is similar to what `__xla_cpu_runtime_OneDnnMatMul` uses. Please refactor it into a function.

File: xla/service/cpu/onednn_matmul_rewriter.cc [hunk 2]
```diff
@@ -522,11 +528,150 @@ class OneDnnMatMulRewriteVisitor : public DfsHloRewriteVisitor {
...
+
+      } else {
+        return DefaultAction(custom_call);
+      }
+    }
+    return DefaultAction(custom_call);
+  }
+
+ private:
+  int intraop_parallelism_;
```
Comment (penpornk) [inline xla/service/cpu/onednn_matmul_rewriter.cc:656]: Nit: We usually treat intra-op as two words.
```suggestion
  int intra_op_parallelism_;
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: xla/service/cpu/ir_emitter.cc [hunk 3]
```diff
@@ -2711,6 +2711,96 @@ Status IrEmitter::HandleOneDnnSoftmax(HloInstruction* custom_call) {
...
+      llvm_ir::EmitAllocaAtFunctionEntry(ptr_array_type, "matmul.args", &b_);
+  llvm::Value* args_life_start =
+      b_.CreateLifetimeStart(args_ptr, b_.getInt64(-1));
+  llvm::Value* args_store = b_.CreateStore(args_val, args_ptr);
+
+  TF_RETURN_IF_ERROR(EmitTargetAddressForOp(custom_call));
+  llvm_ir::IrArray result_array = GetIrArrayFor(custom_call);
+  auto result_stack_alloca = GetAllocaAndEmitMemrefInfo(b_, result_array);
+
+  EmitCallToFunc(runtime::kOneDnnMatMulReorderSymbolName,
```
[the remark this answers is not in the document]
Comment (rfsaliev) [reply xla/service/cpu/ir_emitter.cc]: Done.

File: xla/service/cpu/onednn_matmul_rewriter.cc (hunk 2 above)
Comment (rfsaliev) [reply xla/service/cpu/onednn_matmul_rewriter.cc]: Thank you. I looked to `xla::TfrtCpuClient::eigen_intraop_*` for naming.
Renamed

**Review by penpornk — APPROVED** [review]
Thank you very much for the changes! :)

File: xla/service/cpu/onednn_matmul_rewriter.cc [hunk 4]
```diff
@@ -522,11 +528,150 @@ class OneDnnMatMulRewriteVisitor : public DfsHloRewriteVisitor {
...
+      auto bias_shape =
+          absl::c_count(matmul_config.fused_ops(), OneDnnMatMulConfig::BIAS) > 0
+              ? operands.at(2)->shape()
+              : Shape();
+
+      auto output_shape = custom_call->shape();
+
+#ifndef ENABLE_ONEDNN_OPENMP
+      // set oneDNN cuncurrency settings (which is thread-local)
+      tsl::OneDnnThreadPool::set_onednn_max_threads(intra_op_parallelism_);
```
Comment (penpornk) [inline xla/service/cpu/onednn_matmul_rewriter.cc:615]: The following TF tests failed to build
```
//tensorflow/c:c_api_experimental_test                          FAILED TO BUILD
//tensorflow/c:c_api_function_test                              FAILED TO BUILD
//tensorflow/c:c_api_test_cpu                                   FAILED TO BUILD
//tensorflow/c:c_test                                           FAILED TO BUILD
//tensorflow/c:env_test_cpu                                     FAILED TO BUILD
```

The error message:
```
local_xla/xla/service/cpu/onednn_matmul_rewriter.cc:698:30: error: no type named 'set_onednn_max_threads' in 'tsl::OneDnnThreadPool'
  698 |       tsl::OneDnnThreadPool::set_onednn_max_threads(intra_op_parallelism_);
      |       ~~~~~~~~~~~~~~~~~~~~~~~^
```

**Review by penpornk — CHANGES_REQUESTED** [review]

**Comment by jvstokes** [comment]
@rfsaliev Thank you for the change! Could you please change the reference to   `"@local_tsl//tsl/util:onednn_util_hdrs"` instead to `"@tsl//tsl/util:onednn_util_hdrs"` in xla/service/cpu/BUILD? `local_tsl` doesn't seem to get imported correctly.

**Comment by mdfaijul** [comment]
> @rfsaliev Thank you for the change! Could you please change the reference to `"@local_tsl//tsl/util:onednn_util_hdrs"` instead to `"@tsl//tsl/util:onednn_util_hdrs"` in xla/service/cpu/BUILD? `local_tsl` doesn't seem to get imported correctly.

Thanks! @jvstokes. Corrected accordingly. Please take a look.

**Comment by mdfaijul** [comment]
@jvstokes @penpornk I have noticed that some changes at `third_party/tsl/tsl/util/onednn_threadpool.h` are not included here https://github.com/openxla/xla/pull/8944. As a result there is some build failures there. Could you please take a look.

**Review by jvstokes — APPROVED** [review]

File: xla/service/cpu/onednn_matmul_rewriter.cc (hunk 4 above)
Comment (penpornk) [reply xla/service/cpu/onednn_matmul_rewriter.cc:748]: This error still isn't fixed. I don't think `set_onednn_max_threads` is defined anywhere.

**Review by penpornk — CHANGES_REQUESTED** [review]

**Comment by mdfaijul** [comment]
@penpornk We have defined it inside `third_party/tsl/tsl/util/onednn_threadpool.h`.
https://github.com/openxla/xla/pull/8637/files#diff-166bae5f1b9d2b74f31cd7c9fcbcc6b8ec66389320f2913493e1bd5cac2fc9ad

Could it be because it is not copied to the **tsl** repo https://github.com/google/tsl

**Comment by penpornk** [comment]
> @penpornk We have defined it inside `third_party/tsl/tsl/util/onednn_threadpool.h`. https://github.com/openxla/xla/pull/8637/files#diff-166bae5f1b9d2b74f31cd7c9fcbcc6b8ec66389320f2913493e1bd5cac2fc9ad
> 
> Could it be because it is not copied to the **tsl** repo https://github.com/google/tsl

You're right. I'll copy the changes to TSL manually. Thank you! :)

**Pull request closed without merging**